packages/vite-environment-provider-workerd/src/moduleUtils.ts

```diff
@@ -0,0 +1,71 @@
+import { init as initCjsModuleLexer, parse } from 'cjs-module-lexer';
+import { dirname, resolve } from 'node:path';
+import { readFile, stat } from 'node:fs/promises';
+
+export async function collectModuleInfo(
+  moduleCode: string,
+  moduleFilePath: string,
+): Promise<CjsModuleInfo | EsmModuleInfo> {
+  if (!isCommonJS(moduleCode)) {
+    return { isCommonJS: false };
+  }
+
+  await initCjsModuleLexer();
+
+  const namedExportsSet = new Set<string>();
+
+  const cjsLexerResult = parse(moduleCode);
+  cjsLexerResult.exports.forEach(namedExport =>
+    namedExportsSet.add(namedExport),
+  );
+  for (const reExport of cjsLexerResult.reexports) {
+    const reExportsPath = resolve(dirname(moduleFilePath), reExport);
+
+    const reExportsFileStat = await stat(reExportsPath);
+    if (!reExportsFileStat.isFile) {
+      throw new Error(
+        "Error: Found cjs re-export that doesn't point to a relative path",
+      );
+    }
+
+    const reExportsCode = await readFile(reExportsPath, 'utf8');
+    const reExportsInfo = await collectModuleInfo(reExportsCode, reExportsPath);
+
+    if (reExportsInfo.isCommonJS) {
+      reExportsInfo.namedExports.forEach(namedExport =>
+        namedExportsSet.add(namedExport),
+      );
+    }
+  }
+
+  const namedExports = [...namedExportsSet.values()];
+
+  return {
+    isCommonJS: true,
+    namedExports,
+  };
+}
+
+type CjsModuleInfo = {
+  isCommonJS: true;
+  namedExports: string[];
+};
+
+type EsmModuleInfo = {
+  isCommonJS: false;
+};
+
+function isCommonJS(code: string): boolean {
+  const hasRequireCalls = /\brequire\s*\(\s*['"`][^'"`]+['"`]\s*\)/.test(code);
```

Comment:
This seems OK. I guess there are edge cases if people create their own function named `require` but hopefully that won't happen.

Reply:
yeah 👍 

This whole `isCommonJS` function is pretty brittle I think

Probably a more proper solution would checking the file's `package.json` ([as `vitest-pool-workerd` does](https://github.com/cloudflare/workers-sdk/blob/37dc86ff436aacadd7c0bab6a76293408f4c417f/packages/vitest-pool-workers/src/pool/module-fallback.ts#L103-L132)) plus potentially the file's extension and whatever else node does to decide if a file is cjs or not

but I think that hopefully this simple implementation should be good enough for now and we can improve it later on